galsim/fitswcs.py

```diff
@@ -1343,41 +1340,65 @@ def _radec(self, x, y, color=None):
             assert x.shape == y.shape
             return ra, dec
 
-    def _invert_pv(self, u, v):
-        # Do this in C++ layer for speed.
-        try:
-            len(u)
-        except TypeError:
-            with convert_cpp_errors():
-                return _galsim.InvertPV(u, v, self.pv.ctypes.data)
-        else:
-            invpvu = np.empty(len(u))
-            invpvv = np.empty(len(u))
-            for i in range(len(u)):
-                with convert_cpp_errors():
-                    uu, vv = _galsim.InvertPV(u[i], v[i], self.pv.ctypes.data)
-                invpvu[i] = uu
-                invpvv[i] = vv
-            return invpvu, invpvv
+    def _invert_pv(self, u, v, pv=None, uu=None, vv=None):
+        # Do the inversion via Newton-Raphson iteration.
+        # We're solving u, v = horner2d(u', v', pv) for u' and v' given u and v.
+        # uu and vv are optional initial guesses, if absent, use u and v.
+        if pv is None:
+            pv = self.pv
+        # Assemble horner2d coefs for derivatives
+        _, nx, ny = pv.shape
+        uducoef = (np.arange(nx)[:,None]*pv[0])[1:,:-1]
+        udvcoef = (np.arange(ny)*pv[0])[:-1,1:]
+        vducoef = (np.arange(nx)[:,None]*pv[1])[1:,:-1]
+        vdvcoef = (np.arange(ny)*pv[1])[:-1,1:]
+        if uu is None:
+            uu = u.copy()
+        if vv is None:
+            vv = v.copy()
+
+        # Usually converges in ~3 iterations.
+        for iter in range(10):  # pragma: no branch
+            # Want Jac^-1 . du
+            # du
+            du = horner2d(uu, vv, pv[0], triangle=True) - u
+            dv = horner2d(uu, vv, pv[1], triangle=True) - v
+            # J
+            pvudu = horner2d(uu, vv, uducoef, triangle=True)
+            pvudv = horner2d(uu, vv, udvcoef, triangle=True)
+            pvvdu = horner2d(uu, vv, vducoef, triangle=True)
+            pvvdv = horner2d(uu, vv, vdvcoef, triangle=True)
+            # J^-1 . du
+            det = pvudu*pvvdv - pvudv*pvvdu
+            duu = -(du*pvvdv - dv*pvudv)/det
+            dvv = -(-du*pvvdu + dv*pvudu)/det
+
+            uu += duu
+            vv += dvv
+            # Do at least 2 iterations before spending the time to check for
+            # convergence.
+            if iter > 2 and np.max(np.abs(np.array([duu, dvv]))) < 1e-12:
```

Comment:
I think this does at least 3 iterations?  iter = 0, 1, 2..check.

Reply:
Haha, yes.  Silly me.  The error is in the comment, not the code.  Will fix.

Reply:
Actually, just checked my timing script and `iter > 1` is a bit quicker, so I'll go with that.